File: python/cudaq/shor/qft.py
import math

import cudaq


def apply_qft(kernel: cudaq.Kernel, qubits: list, n: int) -> None:
    """
    Apply the Quantum Fourier Transform on the first n qubits.

    Args:
        kernel: CUDA-Q kernel being built.
        qubits: List of qubit references.
        n: Number of qubits to apply QFT on.
    """
    for i in range(n):
        kernel.h(qubits[i])
        for j in range(i + 1, n):
            angle = math.pi / (2 ** (j - i))
            kernel.cr1(angle, qubits[j], qubits[i])

    for i in range(n // 2):
        kernel.swap(qubits[i], qubits[n - 1 - i])


def apply_inverse_qft(kernel: cudaq.Kernel, qubits: list, n: int) -> None:
    """
    Apply the inverse Quantum Fourier Transform on the first n qubits.

    Args:
        kernel: CUDA-Q kernel being built.
        qubits: List of qubit references.
        n: Number of qubits to apply inverse QFT on.
    """
    for i in range(n // 2):
        kernel.swap(qubits[i], qubits[n - 1 - i])

    for i in range(n - 1, -1, -1):
        for j in range(n - 1, i, -1):
            angle = -math.pi / (2 ** (j - i))
            kernel.cr1(angle, qubits[j], qubits[i])
        kernel.h(qubits[i])

File: rust/qcgpu/src/shor/qft.rs
//! Inverse QFT for the qcgpu OpenCL-based simulator.
//!
//! qcgpu has no built-in QFT, so we implement it manually using the
//! `r(angle: f32)` phase gate. All angles are `f32`, which is sufficient
//! for the small circuit sizes used in the benchmarks (N ≤ 15 with m ≤ 8).

use std::f64::consts::PI;

use qcgpu::gates;
use qcgpu::State;

/// Apply the inverse QFT on the given qubit list.
///
/// qcgpu uses LSB-first indexing: qubit 0 is the rightmost character of the
/// bitstring. The convention here is that `qubits[0]` holds the MSB of the
/// phase after the IQFT swap.
pub fn apply_inverse_qft(state: &mut State, qubits: &[i32]) {
    let n = qubits.len();

    // Inverse-QFT begins by un-swapping the bit order.
    for i in 0..n / 2 {
        state.swap(qubits[i], qubits[n - 1 - i]);
    }

    // Reverse the QFT gates.
    for i in (0..n).rev() {
        for j in ((i + 1)..n).rev() {
            let angle = -(PI / (1u64 << (j - i)) as f64) as f32;
            state.apply_controlled_gate(qubits[j], qubits[i], gates::r(angle));
        }
        state.h(qubits[i]);
    }
}
